extensions/apiconnect/package.json

```diff
@@ -38,14 +38,18 @@
     "url": "https://github.com/strongloop/loopback-next.git",
     "directory": "extensions/apiconnect"
   },
-  "dependencies": {
+  "peerDependencies": {
     "@loopback/core": "^2.9.1",
-    "@loopback/rest": "^5.2.0",
+    "@loopback/rest": "^5.2.0"
```

Comment:
Should `@loopback/core` stay as a regular dependency? We need to clarify the criteria between `dependencies` and `peerDependencies`. I'm thinking about the following:

1. If the dependency needs to be consistent with the application level, it's a peer dependency (always use the one from the application level)
2. Otherwise, it should be a regular dependency (and possibly with a different version than the application level)

Reply:
Another criteria:

If the dependency is the target extension point for this package to contribute extensions, it's a peer dependency. 

Using this rule, `@loopback/core` should be a regular dependency while `@loopback/rest` should be a peer dependency.

Reply:
> - If the dependency needs to be consistent with the application level, it's a peer dependency (always use the one from the application level)
> - If the dependency is the target extension point for this package to contribute extensions, it's a peer dependency.
> - Otherwise, it should be a regular dependency (and possibly with a different version than the application level)

Your criteria make sense to me, however I have a different view on how to interpret them.

In my opinion, there should be only one version of `@loopback/core` installed in any application. Is see the core packages as a cornerstone, a "LoopBack runtime". Among other things, it provides foundation for decorators and manipulation with metadata (`MetadataInspector`, `ParameterDecoratorFactory`, etc.) I am concerned that allowing multiple versions of `@loopback/core` can introduce subtle bugs that will be difficult to troubleshoot, for example when a decorator uses different code for setting metadata than is used by the application to look up that metadata.

Also, if the application developer decides to use their forked version of `@loopback/core` (e.g. to fix a bug), then I want extensions to use the fixed version of core, not the vanilla version from npm.

I would also like to add one more criteria:

- If the dependency is typically installed in target applications, and the extension can support multiple major versions of that dependency at the same time, then the dependency should be a peer dependency.

The use case to consider: where a single version of an extension (e.g. `@loopback/extension-health`) wants to support multiple version of the framework, so that the maintainers can have a single branch and don't have to maintain different branches (major-version release lines) for each framework version. Let's say the extension depends on `@loopback/core@2` and the application depends on `@loopback/core@1`. This is very reasonable, since the only breaking change listed in `@loopback/core@2.0.0` changelog is dropped support for Node.js 8.x. When such extension is installed into the target app, we end up with two copies of `@loopback/core`, where the extension uses a different "LoopBack runtime" than the target application it's installed to.

Let's discuss!

Reply:
That's a valid use case that one extension package can support multiple versions of the LoopBack runtime. But it's hard to ensure that with our CI and release process though.

In addition to the containing application, we probably should also take into consideration that certain LoopBack packages can be used standalone outside a scaffolded LB4 application.

Reply:
> That's a valid use case that one extension package can support multiple versions of the LoopBack runtime. But it's hard to ensure that with our CI and release process though.

I now. Personally, I am fine to keep the support for older major versions on a "best effort" basis with no CI check for now. We can setup a more comprehensive CI solution once there are complaints from users about incompatibilities.

> In addition to the containing application, we probably should also take into consideration that certain LoopBack packages can be used standalone outside a scaffolded LB4 application.

That's a good argument. IIRC, packages that can be used standalone outside of LB4 apps are usually more low level (`@loopback/metadata`, `@loopback/context` etc.) and don't participate in peer dependencies, because they are accessed via higher-level wrappers (`@loopback/core`).

If a LB package is used by a LB4 application that's not scaffolded via `lb4 app`, then I believe the app is still going to depend on `@loopback/core`, therefore my arguments above still apply.

Are you envisioning a project where the developer wants to use an extension depending on `@loobpack/core` but the application does not depend on `@loopback/core`? Would that work at all?

Reply:
> Are you envisioning a project where the developer wants to use an extension depending on @loobpack/core but the application does not depend on @loopback/core? Would that work at all?

No. At minimum, we'll use `@loopback/core`.

Reply:
> No. At minimum, we'll use @loopback/core.

In that case, I think it's reasonable to for extensions to declare `@loopback/core` as a peer dependency, right?

Do you agree to keep the currently proposed direction? May I proceed to rebase this PR on top of latest master (there are going to be many merge conflicts to resolve, so I don't want to do it too many times) and get this PR landed soon?

Reply:
Sure.